src/main/java/com/iexec/worker/compute/ComputeManagerService.java

```diff
@@ -74,8 +81,56 @@ public boolean downloadApp(TaskDescription taskDescription) {
         if (!isDockerType || taskDescription.getAppUri() == null) {
             return false;
         }
+
+        final long pullTimeout = computeImagePullTimeout(taskDescription);
         return dockerService.getClient(taskDescription.getAppUri())
-                .pullImage(taskDescription.getAppUri());
+                .pullImage(taskDescription.getAppUri(), Duration.of(pullTimeout, ChronoUnit.MINUTES));
+    }
+
+    private final Map<Long, Long> alreadyComputedTimeouts = new HashMap<>(5);
+
+    /**
+     * Computes image pull timeout depending on task max time execution.
+     * This should depend on task category (XS, S, M, L, XL).
+     * <p>
+     * Current formula is: {@literal 10 * log(maxExecutionTime * 10)},
+     * with {@literal maxExecutionTime} being expressed in minutes.
+     * The result is rounded to the nearest integer.
+     * <p>
+     * Examples of timeout depending on category
+     * - with default values of 5 for min and 30 for max:
+     * <ul>
+     *     <li>XXS: 10 * log(    20 / 10) =  3 minutes => 5 minutes because of min</li>
+     *     <li>XS : 10 * log(    50 / 10) =  7 minutes</li>
+     *     <li>S  : 10 * log(   200 / 10) = 13 minutes</li>
+     *     <li>M  : 10 * log(   600 / 10) = 18 minutes</li>
+     *     <li>L  : 10 * log(  1800 / 10) = 23 minutes</li>
+     *     <li>XL : 10 * log(  6000 / 10) = 28 minutes</li>
+     *     <li>XXL: 10 * log(100000 / 10) = 40 minutes => 30 minutes because of max</li>
+     * </ul>
+     * If computed timeout duration is lower than {@link DockerRegistryConfiguration#getMinPullTimeout()},
+     * then final timeout duration is {@link DockerRegistryConfiguration#getMinPullTimeout()}.
+     * <br>
+     * If computed timeout duration is greater than {@link DockerRegistryConfiguration#getMaxPullTimeout()},
+     * then final timeout duration is {@link DockerRegistryConfiguration#getMaxPullTimeout()}.
+     * <br>
+     * If {@link DockerRegistryConfiguration#getMinPullTimeout()} is greater than {@link DockerRegistryConfiguration#getMaxPullTimeout()},
+     * then {@link DockerRegistryConfiguration#getMaxPullTimeout()} is the one used.
+     */
+    long computeImagePullTimeout(TaskDescription taskDescription) {
+        final long maxExecutionTime = taskDescription.getMaxExecutionTime() / 60;
+        if (alreadyComputedTimeouts.containsKey(maxExecutionTime)) {
+            return alreadyComputedTimeouts.get(maxExecutionTime);
+        }
+        final long imagePullTimeout = Math.min(
+                Math.max(
+                        Math.round(10.0 * Math.log10(maxExecutionTime / 10.0)),
+                        dockerRegistryConfiguration.getMinPullTimeout()
+                ),
+                dockerRegistryConfiguration.getMaxPullTimeout()
+        );
+        alreadyComputedTimeouts.put(maxExecutionTime, imagePullTimeout);
+        return imagePullTimeout;
```

Comment:
```suggestion
        return alreadyComputedTimeouts.put(maxExecutionTime, imagePullTimeout);
```

Reply:
It won't work as `put` returns the previous value associated with the key.

Reply:
You are right, mixed with .save() :+1: